### PR TITLE
Add an pre-image information into the Enrollment structure

### PR DESCRIPTION
**Pull request by linked0**
A node should reveal an pre-image in advance for effienciency during enrolling as a validator. The pre-image information contains the pre-image as a certain round and the round number, such as: (P, Nx), where P is the pre-image and Nx is: n - (rounds since enrollment), which is described in [the white paper](https://bit.ly/2U6t5O2).

#492

### CODE REVIEW

**Comment by linked0** [comment]
Ready for review

File: source/agora/consensus/EnrollmentManager.d [hunk 1]
```diff
@@ -331,6 +335,10 @@ public class EnrollmentManager
             this.preimages ~= hashFull(this.preimages[i]);
         this.data.random_seed = this.preimages[$-1];
 
+        // Reveal preimage in advance
+        this.data.preimage.round = this.data.cycle_length - PreimagePeriod;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: So the idea is, every 6 rounds the node will release a preimage for the next 6 rounds, right?

Therefore, this data needs to be updated. So shouldn't these two statements be inside of a function that can be called later to update and reveal the preimage?

File: source/agora/consensus/EnrollmentManager.d [hunk 2]
```diff
@@ -61,6 +61,10 @@ public class EnrollmentManager
     /// Enrollment data object
     private Enrollment data;
 
+    /// The round period for revealing a preimage
+    /// It is an hour interval if a block is made in every 10 minutes
+    public static immutable uint PreimagePeriod = 6;
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d:79]: We should make this configurable.

File: source/agora/consensus/data/Enrollment.d [hunk 3]
```diff
@@ -31,6 +31,9 @@ public struct Enrollment
     /// X: random seed, The nth image of random value
     public Hash random_seed;
 
+    /// Preimage information being revealed in advance
+    public PreimageInfo preimage;
```
Comment (Geod24) [inline source/agora/consensus/data/Enrollment.d]: The `PreImageInfo` shouldn't be part of the enrollment, as the enrollment is written on the chain. It should be completely independent. A node can of course send a `PreImageInfo` along its enrollment (we can adapt the API to make this parameter optional), but it should also be able to receive, cache, and propagate PreImageInfo independently.

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/agora/consensus/EnrollmentManager.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: The data will be of course updated during revealing the pre-image after enrollment and the code will be added in another PR. Anyway, I try to apply what you said to the code by making another function for getting an pre-image of a certain round.

File: source/agora/consensus/EnrollmentManager.d (hunk 2 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:79]: I will make another issue now for configuring the period of revealing pre-images. Before the new issue gets completed, we should use that constant value.

File: source/agora/consensus/data/Enrollment.d (hunk 3 above)
Comment (linked0) [reply source/agora/consensus/data/Enrollment.d]: The API endpoint named sendPreimage is planned to be made in another PR. And as you suggested, I will make the `enrollValidator` API function have the second parameter for a pre-image.

**Comment by linked0** [comment]
I have decided to close this PR because it makes unnecessary confusion and the code is not so helpful and it might make the work time unnecessarily longer for the issue. My original intent was to reduce network transmittion by revealing a pre-image in advance during registering as a validator. If such an concept happen to be needed some time, I will apply it to the code. From now I will start implementing the API point called sendPreimage.

**Pull request closed without merging**